### ISSUE TITLE
Exchanging server auth code does not always give id_token

### ISSUE THREAD

**Issue by steven-supersolid** [playgameservices/play-games-plugin-for-unity#1309]
In the Google Play Games config in Unity we have a web client id specified and very minimal resources as we handle achievement id's etc. in our own code.

```
<resources>
  <string name="app_id">xxx</string>
  <string name="package_name">yyy</string>
</resources>

ClientID XXX.apps.googleusercontent.com
```

In the client we retrieve a server auth code as following and successfully get a token in the callback
```
PlayGamesPlatform.Instance.GetServerAuthCode()
```

We pass this token to our server then make a POST request to https://www.googleapis.com/oauth2/v3/token, passing client_id, client_secret, code and grant_type 'authorization_code'.

We get a successful JSON response but sometimes this contains id_token and sometimes access_token but we are unsure why. We require id_token.

**Comment by steven-supersolid** [comment]
Seems related to #1293

**Comment by claywilkinson** [comment]
Currently play game services is switching away from using Google Plus accounts (see http://android-developers.blogspot.com/2016/01/play-games-permissions-are-changing-in.html)  As a result, some players will have a id-token which is provided by G+ and some will not.

You can use the player.id as a key to uniquely identify the player.  The access token returned when exchanging the server auth code is specific to that player as well, so you can make API calls on behalf of the player from your server application.

**Comment by steven-supersolid** [comment]
In our case we only wish to verify the identity of the user on our server, we don't need to make API calls on their behalf.

The blog mentions we can use www.googleapis.com/games/v1/applications/<app_id>/verify/ once we have an access_token. When exchanging a server auth code, will the response consistently contain access_token or will it sometimes be id_token? Is there any documentation available for that endpoint?

BTW we're using the open-source parse-server backend which currently checks a token against the token info endpoint: www.googleapis.com/oauth2/v3/tokeninfo?id_token= but the blog mentions this could be rate limited. This documentation advises to use that endpoint: https://developers.google.com/identity/sign-in/android/backend-auth and this blog still talks about idToken: http://android-developers.blogspot.co.uk/2016/01/using-google-sign-in-with-your-server.html so the situation is a little confusing.

**Comment by claywilkinson** [comment]
Agreed, it is confusing, and the game services team is working to make it more clear :)

If you really need the ID token, you can force players to make sure they have a Plus account (and consent to using their identity by enabling the RequireGooglePlus attribute on the builder.

**Comment by claywilkinson** [comment]
The documentation for games/v1/application/verify is at https://developers.google.com/games/services/web/api/applications/verify

**Comment by steven-supersolid** [comment]
Thanks for all the info. Closing now as not an issue with the plugin

**Comment by steven-supersolid** [comment]
Just an update on this.

I reduced the scopes on the login dialog to just profile and email, but some users have a new gamer id of the form g1234 instead of 1234. Assuming the latter is a Plus id and by removing the RequireGooglePlus attribute I was no longer forcing users to have a Plus id.

I switched to the access_token checking endpoint: GET https://www.googleapis.com/games/v1/applications/applicationId/verify and this worked. I didn't see any documentation but the query string parameter should be access_token in this case.

The endpoint also works for numberical id's so I'll be using that from now on. I think I may be able to reduce scopes further and still get access_token but still need to test to be sure.
